### ISSUE TITLE
Generate report per test?

### ISSUE THREAD

**Issue by ffMathy** [lucaslorentz/minicover#130]
This is something Coverlet can't do, so I'm curious if this is something Minicover is interested in doing.

See coverlet-coverage/coverlet#973 for an explanation on why I'd like something like this.

**Comment by ffMathy** [comment]
More context (from the above issue):

> I want to build a CLI that internally runs tests for you serially (non-parallel) that have been affected by your GIT changes, based on previous coverage.
> 
> That's why I want to be able to see what tests ran through a specific line of code.
> 
> See the examples given in my other issue here: dotnet/aspnetcore#25194

**Comment by lucaslorentz** [comment]
The approach we take is pretty close to what was described in this comment: https://github.com/coverlet-coverage/coverlet/issues/973#issuecomment-715211965
We use AsyncLocal to store a "method context", but we don't use the call stack to find the test method.

We do 2 kinds of instrumentation:
1 - when a method enters we try to find an existing "method context" in AsyncLocal, and create a new one if no context is found
2 - before every IL instruction in a method, we register a hit and relate it to the current "method context"

For source code (argument --sources), we apply instrumentation 1 and 2.
For test files (argument --tests), we apply only instrumentation 1.

So, basically, every hit is registered in the context of the outermost instrumented method. Which happens to be the test method when you inform your test files. As you can see in this HTML report:
![image](https://user-images.githubusercontent.com/2268284/97356676-e457ff80-1898-11eb-85d6-047d65fe673b.png)

The bad news is that we only output that information in the HTML report.
So in order to access it, we would need to either extend one of the existing reports or create a new report.

Let me know if I can help you further.

**Comment by ffMathy** [comment]
Very cool @lucaslorentz! Would you be interested in merging a PR where a brand new reporter is added, that just serializes the full report as-is in JSON format?

Coverlet for instance does that.

**Comment by lucaslorentz** [comment]
Yes, sure, I would be glad to.

**Comment by ffMathy** [comment]
While I'm at it, is there anything stopping us from rewriting this in NET Standard?

I get that the CLI needs to be NET Core, but I think everything else should be extracted out into its own project, and exposed as an individual NuGet package.

My tool could then refer that package instead, and that package could be written in Standard.

What do you think of this? Because then I'll do that too 😊

**Comment by lucaslorentz** [comment]
Nope, just need to split into smaller projects.

But keep the CLI inside Minicover package be backwards compatible.

Btw, took some hours to debug it, and will add soon a possible fix to:
https://github.com/lucaslorentz/minicover/issues/103
https://github.com/lucaslorentz/minicover/issues/119
https://github.com/lucaslorentz/minicover/issues/126
https://github.com/lucaslorentz/minicover/issues/127

**Comment by lucaslorentz** [comment]
Packages are there:
https://www.nuget.org/packages/MiniCover.Core/
https://www.nuget.org/packages/MiniCover.Reports/

Let me know if they work for you.

**Comment by ffMathy** [comment]
I'll definitely give it a shot once #136 is fixed! Thank you so much.
